### DISCUSSION TITLE
LevelEdit general update

### DISCUSSION THREAD

**Discussion by koanxd** [CorsixTH/CorsixTH#3402] [Research Room]
I'm planning to do extensive changes to the level editor, primarily adding missing features such as epidemics and the remaining awards and trophies, also fixing some minor issues. Refactoring is badly needed as well.

Since nobody else is currently working on it, I will do several commits in a single pull request when I'm finished.

Please tell me if there is anything to discuss, or if you have suggestions.

I have some questions myself: will all the level variables eventually be added into the game? If yes, I might also add them to the editor, but with a note that they may not be implemented.

I am especially interested in the disease ratios, which are to be saved in 'visuals' and 'nonvisuals'. Would the result be the same if every number is multiplied by 10, as it is the same ratio? Just to figure out how it would work.

**Comment by Grimdoc** [comment]
Suggestion: Make an option to place the entire helipad, so that the user
don't have to place all the tiles himself.
We have seen maps where the users have been unable to get helicopters
(because they placed some tiles wrong, or used wrong tiles).

**Comment by koanxd** [comment]
Sorry, I wasn't talking about the map editor, I wouldn't want to mess with that yet. I meant the one for editing the .level textfiles.

**Comment by Glimmer-Man** [comment]
1.
In emergency setting - option semi-random - there is issue about two emergency in one month. Look at #422. So it will be great, that wouldn'n be possile to set result <31 days.
1. Population number - when I set 0 to "change per month" in last row, I still see "patients count will increase infinitely after month", I must add and remove row if I want that disappeared.
2. Option to disable earthquakes.

**Comment by Lego3** [comment]
Great. :-) I have done a little refactoring, but more is needed.

**Comment by koanxd** [comment]
Thanks, I will address all your points. Regarding semi-random, I think the best way is to display a warning that it may cause issues if the values are set too low.

**Comment by admdly** [comment]
Potentially unrelated, but I can see an issue with the distribution of LevelEdit as 'first-party' app given it requires Java. I know it has been suggested that at some point LevelEdit will ship with the standard distribution, but as this will require Java on the users' machine I can't see this being viable (and certainly not something I'd be interested in for OS X).

I'd much prefer to see this be coded in a way which does not require something like Java (nothing against the original developer, as someone who also 'has' to use Java). I just can't, personally, justify pulling in Java as a requirement for a standard build of CorsixTH.

That said, I also don't use Lua on a daily basis either. This project is my entry to Lua per se. However, it does not require any third party frameworks on the user's computer.

Usually, I'd be interested in porting the LevelEdit app to be more cross platform (Python or C++/Lua) but I'm extremely busy at the moment, so I'm not complaining. My overall comment is that I can't see justifying requiring Java just for LevelEdit for a standard distribution of CorsixTH (and it is not something I will ever, personally, support with OS X builds).

**Comment by TheCycoONE** [comment]
@admdly the LevelEdit is already a "first-party" app, shipped with the Windows releases.

There is plans to merge it with MapEdit (whose own future is a little up in the air right now), but no one is working on it.

**Comment by koanxd** [comment]
I realize that it's not pretty, and I'm not a fan of java eiter. So I don't mind if or how this ends up being used..

**Comment by TheCycoONE** [comment]
@koanxd what you propose is worthwhile.  There is a lot of work to be done before LevelEdit becomes redundant, so I strongly suspect it will be shipped with the next version.

**Comment by koanxd** [comment]
I read that leveledit will not be making it to 0.50, which is reasonable. I know it has been too long, but I haven't abandoned this project. Is a replacement planned? Otherwise, I will continue working on this, even though motivation is an issue since I am rewriting the whole thing.

**Comment by TheCycoONE** [comment]
It would be nice if the level edit capabilities were part of the map edit functionality which is being integrated into the game, but no one is working on that right now and there's no reason to think it will happen soon.

If you would like to take ownership of the LevelEdit Java tool in the mean time I would prefer that it become it's own project under the CorsixTH organization with it's own independent releases, similar to the CorsixTH-Graphics project.

If that is something you're interested in, I could set that up for you.

**Comment by koanxd** [comment]
Sounds good to me.

**Comment by Lego3** [comment]
Has anything happened here?

**Comment by koanxd** [comment]
Not that I know of. It's been in a state of "almost done" for a while now, but it wouldn't make sense to do a commit due to the extensive and still ongoing changes. Sorry about that. Most notably, language can now be changed during runtime, and the writing process will now leave the file as it was, except where the user made changes. I'm also getting rid of the copypaste-code, in the hopes of creating something that is maintainable.. which it still might not be, but at least it will be documented.

As I said before, if someone else wants to step in, please go ahead. For me, this started as a personal learning project, but yes I do still work on it occasionally.

**Comment by lloyd-lejack** [comment]
Is leveledit still under dev or snowblind has ceased to work on it? (As it have been two years since the last post here)
Im interested since its a real hell to edit manually everything.

**Comment by TheCycoONE** [comment]
I haven't heard anything since that last update two years ago.

**Comment by lloyd-lejack** [comment]
Then probably dead, as snowblind was active when i made my acc at the forums but its no longer active.
Really bad, as LevelEdit just makes everything easier.

**Comment by koanxd** [comment]
To be honest, after a while I was too embarrassed to post here since I was never able to show any good work. I underestimated the amount of effort and got frustrated.

Since there is still interest, I'm giving it another try. If I don't reply by the end of next month I will have quit for good, though.

(In case anyone wants to see the latest source, let me know. I just don't think it would be of much use.)

**Comment by Alberth289346** [comment]
Try to reduce the scope of the patch, better a small finished patch, than a complete solution that is never created

**Comment by lloyd-lejack** [comment]
Glad to hear (actually read haha) that you're giving another try to this tool, snowblind. Hope i see some progress. Don't give up on it, It's a really helpful tool man!
And as Alberth said, it's better a working small patch than an unfinished full version.

**Comment by lloyd-lejack** [comment]
February is almost over and we have no news from snowblind. Does this mean he may have ceased developing LevelEdit? (Since he said if he didn't reply by the end of this month he would have quit for good)

**Comment by koanxd** [comment]
It just means I'm the type of person to draw things out until the last minute. I plan to upload on the 28th or the day after. I'll be taking Alberth's advice to reduce the scope as well.

**Comment by lloyd-lejack** [comment]
Oh, i see. Sorry if my message sounded a bit rude. :p Looking forward for your update.

**Comment by koanxd** [comment]
I've uploaded a new version under a separate repository: https://github.com/koanxd/corsixTH-LevelEdit

Note that this commit is based on an out-dated version, so the most recent LevelEdit changes by Alberth are not included.

I am also not familiar with build tools, so I have included an executable .jar file.

A full changelog is included as a commit message. In hindsight, this patch should have been done in much smaller steps. I am sure there are new bugs introduced in this version, so further testing is needed before using it. Regardless, I hope something can be gained from this eventually.

**Comment by lloyd-lejack** [comment]
https://prnt.sc/indyan Ok, i can't save nor load files, not good.

**Comment by koanxd** [comment]
I don't know what could be causing it. I suspect it may be related to Windows permissions. Can anyone else reproduce the error?

I've opened an issue here: https://github.com/koanxd/corsixTH-LevelEdit/issues/1

**Comment by lloyd-lejack** [comment]
Personally i haven't tried to replicate this error since it first happened. I was waiting for someone besides you to do so. But since no one has I'll probably retry tomorrow.

**Comment by lewri** [comment]
Do we know what has been done with LevelEdit over the past 3 years? I don't think anything. Last time I think @Lloydjara was building the level file manually.

**Comment by Alberth289346** [comment]
Yes we know, or rather, git knows. Ask for a log in the `LevelEdit` directory.

I started a Lua version recently, but the number of settings is huge, to the point that I don't want to code everything. Instead the data that describes the settings and build a GUI for editing should be read from a file.

**Comment by lewri** [comment]
@Alberth289346 should I tie #2399 to this ticket or shall we close this?

**Comment by Alberth289346** [comment]
I am not sure.

Arguably the Java version is much more available to users than my Lua version. My code is stuck on hiding support of the widgets, and there isn't much movement in the generic Window classes code.

On the other hand, the Java version works, but it's not nice. I hacked it a bit to make it follow Java rules about comparing values, but it also has loads of copy/pasted code.

Personally I don't mind either way.

For the project, I would guess the Lua version is more useful, but then the window classes need to move first.

**Comment by lewri** [comment]
I would be inclined that a built in application than a separate one is preferable, and also means we don't have to support Java as well as c++/lua.
